### PR TITLE
feat: implement trezorlib

### PR DESCRIPTION
**Pull request by sabotagebeats**
Build ape-trezor implementation with Trezorlib.

### CODE REVIEW

File: ape_trezor/accounts.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            txn.nonce,
+            txn.gas_price,
+            txn.gas_limit,
+            txn.receiver,
+            txn.value,
+            txn.data,
+            txn.chain_id,
+            # tx_type,
+        )
+        txn.signature = signature
```
Comment (fubuloubu) [inline ape_trezor/accounts.py:83]: I think this might be like this
```suggestion
        txn.signature = signature["signature"]
```

**Review by fubuloubu — APPROVED** [review]
Run `black` and other linting steps

**Review by sabotagebeats — COMMENTED** [review]
*tap tap* is this thing on? 😅

File: ape_trezor/_cli.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+@cli.command(
+    short_help="Remove an Trezor account from your Ape configuration. \
+    (The account will not be deleted from the Trezor hardware device)"
+)
+@click.argument("alias")
+def delete(alias):
+    path = container.data_folder.joinpath(f"{alias}.json")
+    try:
+        path.unlink()
+    except Exception as e:
```
[the remark this answers is not in the document]
Comment (fubuloubu) [reply ape_trezor/_cli.py:88]: Sure!

File: setup.py [hunk 3]
```diff
@@ -63,7 +60,14 @@
...
         "importlib-metadata ; python_version<'3.8'",
+        "eth-ape>=0.1.0a17,<0.2.0",
+        "trezor>=0.12.2",
     ],  # NOTE: Add 3rd party libraries here
+    entry_points={
+        "ape_cli_subcommands": [
+            "ape_trezor=ape_trezor._cli:cli",
+        ],
+    },
     python_requires=">=3.6,<4",
```
[the remark this answers is not in the document]
Comment (fubuloubu) [reply setup.py:71]: Yes!

File: ape_trezor/accounts.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        if msg.version != b"E": 
+            return None
+        # TODO: trezor does not support eip712 yet, it only supports eip 191 personal_sign
+        
+        signature = ethereum.sign_message(self.client, parse_hdpath(self.hdpath), msg.body)
+        messagehash = _hash_eip191_message(msg)
+        r = signature['signature'][0:32]
+        s = signature['signature'][32:64]
+        v = signature['signature'][64]
+        return SignedMessage(messagehash,r,s,v,signature['signature'])
```
[the remark this answers is not in the document]
Comment (fubuloubu) [reply ape_trezor/accounts.py]: Yes, this is just a matter of casting to `HexBytes`. Do `from hexbytes import HexBytes` and then `HexBytes(messagehash)`

**Comment by sabotagebeats** [comment]
```
In [2]: token = a.deploy(project.Token,"hello","WRLD",18,100000*10**18)
Passphrase required: 
Confirm your passphrase: 
Please confirm action on your Trezor device.
> /home/sabotage/.pyenv/versions/apeworx/lib/python3.8/site-packages/ape_trezor-0.1.0a1.dev16+gc6410f6.d20210805-py3.8.egg/ape_trezor/accounts.py(82)sign_transaction()
-> return  TransactionSignature(bytes(vrs[0]), vrs[1], vrs[2])
(Pdb) c
> /home/sabotage/.pyenv/versions/apeworx/lib/python3.8/site-packages/eth_account/_utils/legacy_transactions.py(67)encode_transaction()
-> return rlp.encode(signed_transaction)
(Pdb) c
---------------------------------------------------------------------------
SerializationError                        Traceback (most recent call last)
~/.pyenv/versions/apeworx/lib/python3.8/site-packages/rlp/sedes/lists.py in serialize(self, obj)
     75             try:
---> 76                 yield sedes.serialize(element)
     77             except SerializationError as e:

~/.pyenv/versions/apeworx/lib/python3.8/site-packages/rlp/sedes/big_endian_int.py in serialize(self, obj)
     20         if isinstance(obj, bool) or not isinstance(obj, int):
---> 21             raise SerializationError('Can only serialize integers', obj)
     22         if self.l is not None and obj >= 256**self.l:

SerializationError: Can only serialize integers

During handling of the above exception, another exception occurred:

ListSerializationError                    Traceback (most recent call last)
~/.pyenv/versions/apeworx/lib/python3.8/site-packages/rlp/sedes/serializable.py in serialize(cls, obj)
    265         try:
--> 266             return cls._meta.sedes.serialize(obj)
    267         except ListSerializationError as e:

~/.pyenv/versions/apeworx/lib/python3.8/site-packages/eth_utils/functional.py in inner(*args, **kwargs)
     44         def inner(*args, **kwargs) -> T:  # type: ignore
---> 45             return callback(fn(*args, **kwargs))
     46 

~/.pyenv/versions/apeworx/lib/python3.8/site-packages/rlp/sedes/lists.py in serialize(self, obj)
     77             except SerializationError as e:
---> 78                 raise ListSerializationError(obj=obj, element_exception=e, index=index)
     79 

ListSerializationError: Serialization failed because of element at index 6 ("Can only serialize integers")

During handling of the above exception, another exception occurred:

ObjectSerializationError                  Traceback (most recent call last)
<ipython-input-2-3e257b886a75> in <module>
----> 1 token = a.deploy(project.Token,"hello","WRLD",18,100000*10**18)

~/ape/ape/src/ape/api/accounts.py in deploy(self, contract_type, *args, **kwargs)
    117         txn = c(*args, **kwargs)
    118         txn.sender = self.address
--> 119         receipt = self.call(txn)
    120 
    121         if not receipt.contract_address:

~/ape/ape/src/ape/api/accounts.py in call(self, txn, send_everything)
     79             raise Exception("User didn't sign!")
     80 
---> 81         return self.provider.send_transaction(txn)
     82 
     83     @cached_property

~/ape/ape-infura/ape_infura/providers.py in send_transaction(self, txn)
     56 
     57     def send_transaction(self, txn: TransactionAPI) -> ReceiptAPI:
---> 58         txn_hash = self._web3.eth.send_raw_transaction(txn.encode())
     59         return self.get_transaction(txn_hash.hex())
     60 

~/ape/ape/src/ape_ethereum/ecosystem.py in encode(self)
     57         signature = (self.signature.v, self.signature.r, self.signature.s)
     58 
---> 59         signed_txn = encode_transaction(unsigned_txn, signature)
     60 
     61         if self.sender and EthAccount.recover_transaction(signed_txn) != self.sender:

~/.pyenv/versions/apeworx/lib/python3.8/site-packages/eth_account/_utils/legacy_transactions.py in encode_transaction(unsigned_transaction, vrs)
     65     signed_transaction = Transaction(v=v, r=r, s=s, **chain_naive_transaction)
     66     breakpoint()
---> 67     return rlp.encode(signed_transaction)
     68 
     69 

~/.pyenv/versions/apeworx/lib/python3.8/site-packages/rlp/codec.py in encode(obj, sedes, infer_serializer, cache)
    118         item = sedes.serialize(obj)
    119     elif infer_serializer:
--> 120         item = infer_sedes(obj).serialize(obj)
    121     else:
    122         item = obj

~/.pyenv/versions/apeworx/lib/python3.8/site-packages/rlp/sedes/serializable.py in serialize(cls, obj)
    266             return cls._meta.sedes.serialize(obj)
    267         except ListSerializationError as e:
--> 268             raise ObjectSerializationError(obj=obj, sedes=cls, list_exception=e)
    269 
    270     @classmethod

ObjectSerializationError: Serialization failed because of field v ("Can only serialize integers")
```

**Comment by sabotagebeats** [comment]
If I convert v to int, then it fails with the same error on r, but if I try to convert r to int then it gives me that I can't convert bytes to int. 
I also gave this a shot converting to hexbytes and got the same error..

**Comment by fubuloubu** [comment]
I believe the issue might actually have to do with this code:

https://github.com/ApeWorX/ape/blob/main/src/ape/types/signatures.py#L9
https://github.com/ApeWorX/ape/blob/main/src/ape_ethereum/ecosystem.py#L57

**Comment by sabotagebeats** [comment]
https://github.com/ApeWorX/ape/blob/main/src/ape_ethereum/ecosystem.py#L61
it fails on line 61 which calls this: 
https://github.com/ethereum/eth-account/blob/5558be9ce3fef8fe396c45bb744e377c7af5333d/eth_account/_utils/legacy_transactions.py#L55

**Comment by sabotagebeats** [comment]
I also tried this
```
(Pdb) from rlp.sedes import Binary, binary
(Pdb) signature = (self.signature.v, Binary(self.signature.r), Binary(self.signature.s))
```
but I still get the similar error:
`*** rlp.exceptions.ObjectSerializationError: Serialization failed because of field r ("Can only serialize integers")`

**Comment by fubuloubu** [comment]
> https://github.com/ApeWorX/ape/blob/main/src/ape_ethereum/ecosystem.py#L61
> it fails on line 61 which calls this:
> https://github.com/ethereum/eth-account/blob/master/eth_account/_utils/legacy_transactions.py#L55

So I think this line needs to get modified:
https://github.com/ApeWorX/ape/blob/main/src/ape_ethereum/ecosystem.py#L57

using `eth_utils.to_int` like this:

```python
signature = (self.signature.v, to_int(self.signature.r), to_int(self.signature.s))
```

**Comment by sabotagebeats** [comment]
@fubuloubu thanks I'll give that a try!

**Comment by fubuloubu** [comment]
> @fubuloubu thanks I'll give that a try!

https://github.com/fubuloubu/ape/tree/fix/integer-signatures

**Comment by sabotagebeats** [comment]
@fubuloubu this works and I was able to push the contract to ropsten

**Comment by fubuloubu** [comment]
Wait on this: https://github.com/ApeWorX/ape/pull/124

File: README.md [hunk 5]
```diff
@@ -28,7 +28,16 @@ python3 setup.py install
 
 ## Quick Usage
 
-TODO: Describe library overview in code
+```bash
+ape trezor add $account_name
```
Comment (fubuloubu) [inline README.md:32]:
```suggestion
ape trezor add [PICK AN ALIAS]
```

File: README.md [hunk 6]
```diff
@@ -28,7 +28,16 @@ python3 setup.py install
...
+```
+you can now load the account like any other account in Ape console and then use it to sign transactions.
+
+```bash
+ape trezor sign-message $account_name "hello world"
```
Comment (fubuloubu) [inline README.md:37]:
```suggestion
ape trezor sign-message [YOUR TREZOR ALIAS] "hello world"
```

**Review by fubuloubu — COMMENTED** [review]
Just some suggestions to make it clear that you need to fill it in, and it's not valid bash

**Comment by fubuloubu** [comment]
> @fubuloubu any final requests for change before merging? Should I get tests working first? That will require emulator in CI i think.

Fix up the linting and type-check CIs first, but yeah for tests just add a simple test that the `ape trezor remove` function works (place any `.json` file in that dir). Everything should pass CI after you get that done, and I'll approve and we can merge.

**Comment by fubuloubu** [comment]
Checkout: https://click.palletsprojects.com/en/8.0.x/testing/?highlight=testing

**Comment by sabotagebeats** [comment]
> Checkout: https://click.palletsprojects.com/en/8.0.x/testing/

Thanks i'll refactor this with that info

**Comment by fubuloubu** [comment]
> > Checkout: https://click.palletsprojects.com/en/8.0.x/testing/
> 
> Thanks i'll refactor this with that info

github CI is down, so... hahaha

**Comment by sabotagebeats** [comment]
this is what I'm getting in my local pytest. I'm thinking this is because it isn't able to instantiate the local accounts container in pytest then fails when trying to import it. 

```
tests/test_integration.py:2: in <module>
    from ape_trezor._cli import delete
../../.pyenv/versions/3.8.10/envs/apeworx/lib/python3.8/site-packages/ape_trezor-0.1.0a1.dev19+g64b9d20.d20210809-py3.8.egg/ape_trezor/_cli.py:13: in <module>
    container = accounts.containers["trezor"]
E   KeyError: 'trezor'
```

**Comment by fubuloubu** [comment]
> this is what I'm getting in my local pytest. I'm thinking this is because it isn't able to instantiate the local accounts container in pytest then fails when trying to import it.
> 
> ```
> tests/test_integration.py:2: in <module>
>     from ape_trezor._cli import delete
> ../../.pyenv/versions/3.8.10/envs/apeworx/lib/python3.8/site-packages/ape_trezor-0.1.0a1.dev19+g64b9d20.d20210809-py3.8.egg/ape_trezor/_cli.py:13: in <module>
>     container = accounts.containers["trezor"]
> E   KeyError: 'trezor'
> ```

That's awkward, let's work through this on the call tomorrow

**Comment by sabotagebeats** [comment]
@fubuloubu do you know what the type check is failing on? 

```
$ mypy -p ape_trezor
ape_trezor/accounts.py:62: error: Too many arguments for "MessageSignature"
ape_trezor/accounts.py:78: error: Too many arguments for "TransactionSignature"
Found 2 errors in 1 file (checked 4 source files)
```

but these I believe have the correct amount of arguments as required?

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**